### PR TITLE
API server

### PR DESCRIPTION
**Pull request by jackzampolin**
cc @anilCSE @akhilkumarpilli 

[See the list of routes as well as comments](https://github.com/cosmos/relayer/pull/323/files#diff-89cdc96cd44fc55cb509f62f12a3dc4173c8b9dfd81222cdff65f8544577920bR73)

ref: #52

### CODE REVIEW

**Comment by akhilkumarpilli** [comment]
@jackzampolin, Can you share me checklist of api endpoints here so it would be easy for us to implement them.

**Comment by jackzampolin** [comment]
@akhilkumarpilli Done! I've got them all stubbed out

**Comment by anilcse** [comment]
@akhilkumarpilli can you fix the conflicts?

File: helpers/light.go
Comment (odeke-em) [inline helpers/light.go]: Perhaps this code needs to checks if height <= 0, otherwise if line 14 slips through with opts[0] being say -10, line 30 will be reached.

File: helpers/light.go
Comment (odeke-em) [inline helpers/light.go]: Please fold this into
```go
return chain.GetLatestLightHeader()
```

File: helpers/light.go
Comment (odeke-em) [inline helpers/light.go]: This code will never be executed, given that you have the default case :-)

File: helpers/query.go
Comment (odeke-em) [inline helpers/query.go]: We can reverse this and reduce on nesting with an early return
```go
if len(dts.DenomTraces) == 0 {
      return coins, nil
}

var out sdk.Coins
for _, c := range coins {
     ...
}
return out, nil
```

File: helpers/query.go
Comment (odeke-em) [inline helpers/query.go]: Ditto here, please check `if height <= 0` otherwise opts[0] == "-10" will slip through the cracks.

File: helpers/query.go
Comment (odeke-em) [inline helpers/query.go]: Should this be `if height < 0` ?

File: helpers/query.go
Comment (odeke-em) [inline helpers/query.go]: Perhaps `return chain.QueryTxs(relayer.MustGetHeight(h.GetHeight()), int(offset), int(limit), events)`?

File: relayer/query.go
Comment (odeke-em) [inline relayer/query.go:52]: Thank you for removing this!

File: relayer/query.go
Comment (odeke-em) [inline relayer/query.go]:
```go
addr, err := sdk.AccAddressFromBech32(address)
done()
if err != nil {
    return nil, err
}
```

File: relayer/strategies.go
Comment (odeke-em) [inline relayer/strategies.go]: Could we perhaps perform a check on the error to confirm if it is really started or if something else occured?

File: relayer/strategies.go
Comment (odeke-em) [inline relayer/strategies.go]: Ditto here about checking the error if it is related to starting.

File: helpers/api.go
Comment (odeke-em) [inline helpers/api.go]: If you do a `strings.TrimSpace(r.URL.Query().Get("height"))` here and elsewhere, you can conclusively determine if empty strings were sent and can perform the chain.QueryLatestHeight() :)

**Review by odeke-em — COMMENTED** [review]
Thank you @akhilkumarpilli @jackzampolin for this change! Just a drive-by review for security and nits, I've added some comments, please take a look.

**Comment by colin-axner** [comment]
I will review this sometime this week, just haven't gotten a chance yet since there are a lot of changes

File: cmd/api.go
Comment (colin-axner) [inline cmd/api.go]: client/connection/channel commands change the config during execution so we would want to make sure we are aware of that when addressing this TODO

File: cmd/api.go
Comment (colin-axner) [inline cmd/api.go]: We should open issues for every TODO

File: helpers/query.go
Comment (colin-axner) [inline helpers/query.go]:
```suggestion
				out = append(out, sdk.NewCoin(d.IBCDenom(), c.Amount))
```
this will panic if provided the full denom path

File: helpers/api.go
Comment (colin-axner) [inline helpers/api.go]: any reason not to check the error here?

File: helpers/api.go
Comment (colin-axner) [inline helpers/api.go]: interesting we don't need a w.Write call?

**Review by colin-axner — COMMENTED** [review]
Overall LGTM. I did my best to review this, but it is a lot of diffs, and most of it looks like a refactor of old function structuring. Just left a couple questions on the rest response writing

Would appreciate another reviewer to take a look at this, but it does only touch client code so it's ok if bugs are not caught during review. 

@akhilkumarpilli have you tested this manually for a couple of the rest routes? Could you add a quick document on how to interact with this?

**Comment by colin-axner** [comment]
also waiting on review from @fedekunze

File: cmd/api.go
Comment (akhilkumarpilli) [reply cmd/api.go]: @colin-axner , do we need to remove this TODO since we thought of not implementing APIS for tx commands?

File: helpers/api.go
Comment (akhilkumarpilli) [reply helpers/api.go]: We can use w.Write instead of json.NewEncoder but we need to convert errorResponse struct to bytes. So, I just used json.NewEncoder.

**Comment by akhilkumarpilli** [comment]
> @akhilkumarpilli have you tested this manually for a couple of the rest routes? Could you add a quick document on how to interact with this?

@colin-axner , We will setup swagger docs for rest APIs in a separate PR.

File: cmd/api.go
Comment (colin-axner) [reply cmd/api.go]: lets remove

**Comment by colin-axner** [comment]
Adding swagger docs later sounds fine, but I think someone should manually test to at least ensure one route works. Is there a reason why this cannot be done? Is it because we'd need some third party code to submit the rest requests?

**Comment by akhilkumarpilli** [comment]
Not a problem, I will add a short document regarding how to start rest server and interact with routes. We thought swagger docs might be helpful since we can test routes directly from it. We will add swagger docs later in other PR and push a quick document now.

**Comment by colin-axner** [comment]
sounds perfect, thanks @akhilkumarpilli

File: cmd/api.go
Comment (fedekunze) [inline cmd/api.go]: these are really hard to read, what's the need for `fmt.Sprintf`? I think the value of having the endpoints as constants is less than using a plain string

File: cmd/keys.go
Comment (fedekunze) [inline cmd/keys.go]: the sdk has a constant for this

**Review by fedekunze — APPROVED** [review]
utACK, although I think this API could be further refactored / cleaned up. Can you also fix the alerts from lgtm-com?

File: cmd/api.go
Comment (jackzampolin) [inline cmd/api.go:605]: Does this work? That would a step change in the relayer capabilities.

Comment (akhilkumarpilli) [reply cmd/api.go:605]: We need to do some rework for this api. We can implement all TODOs related to `PostRelayerListenHandler` in separate PR.

**Comment by akhilkumarpilli** [comment]
>Can you also fix the alerts from lgtm-com?

I tried to fix lgtm-com alerts but I am unable to fix them.

File: cmd/api.go
Comment (colin-axner) [reply cmd/api.go:605]: could you open an issue to track this work?

**Comment by colin-axner** [comment]
> I tried to fix lgtm-com alerts but I am unable to fix them.

The incorrect conversion can be fixed by changing the upstream offset value to int instead of uint64 or by adding a if statement before doing QueryTxs action. 

The uncontrolled data path I think we can just suppress those errors (click on the hide button on lgtm to see how to suppress). The suppression won't take place until after this is merged, but we should still add it

File: cmd/api.go
Comment (colin-axner) [inline cmd/api.go:83]: Does this {name} automatically get interpreted as an argument?

Comment (akhilkumarpilli) [reply cmd/api.go:83]: Yes, it will be argument

File: helpers/query.go
Comment (colin-axner) [inline helpers/query.go:80]: I think this should be MaxInt32? right? I believe int is 32 bytes though apparently not an alias for int32

**Review by colin-axner — APPROVED** [review]
utACK, I think we can merge this in so it starts getting some testing, see my MaxInt32 comment first though since it seems like a small fix

File: helpers/query.go
Comment (akhilkumarpilli) [reply helpers/query.go:80]: Okay, but I think int is atleast 32 bytes. It's just depends on platform and can be 64 bytes I guess. Is it right?

Comment (colin-axner) [reply helpers/query.go:80]: that sounds correct. Let's leave as is then

**Pull request merged**
